### DISCUSSION TITLE
User Profile - Make attribute "readonly" based on criteria (plaintext and regex)

### DISCUSSION THREAD

**Discussion by mpalavrov** [keycloak/keycloak#25621] [Ideas]
Hi Keycloak team,

We have a request by a customer of ours, which is actually a very good suggestion to improving the User Profile attributes.
The request is to be able to make an attribute "readonly" based on criteria (could be plain text or regex).

This should be achieved by evaluating only the already submitted attributes in the User Profile, not fro attributes, currently submitted during registration or user update.

It can be achieved using annotation, similar to:
![image](https://github.com/keycloak/keycloak/assets/33352511/60aaa2e7-b312-4359-ab18-4ca1483de9c1)

or for a checkbox, where this feature is actually requested for (if the terms and conditions are already accepted, to be readonly):
![image](https://github.com/keycloak/keycloak/assets/33352511/0c813c53-78d0-4f49-8fec-fdc44132d349)

**Comment by mpalavrov** [comment]
I think that the same can be done for hiding/showing (enable) the attributes in addition, using second annotation "hidden-criteria" for example

**Comment by ssilvert** [comment]
I agree that this kind of functionality is needed.  And you would also want the same kind of thing for required fields where "this field is required based on the value of some other field".

@pedroigor @edewit WDYT?

**Reply by jbman** [reply]
This suggestion is not yet about inter-dependencies of attributes. 
Make an attribute required based on some other field is also nice and I think it could also be enhanced later. We also have customers who want to have even more complicated logic, like defining selectable values based on the value of an other attribute. So maybe this could be discussed in a separate thread?

The question here is how the feature could be realized. Annotation based is only one option, but there could also be a `read-only-if value-matches` property on the attribute and a `hidden-if-value-matches`.
These two features allow to make final values of an attribute possible which can not be changed by the user, once they are stored. 

Example use cases: 
- providing a customer number which should not be changed once provided.
- confirming that a privacy policy or terms of service was accepted
- starting an irrevocable deletion request

**Comment by pedroigor** [comment]
@mpalavrov Is it about validating if an attribute is required based on a regex against the attribute value? Not sure if I'm following the example using an annotation ...

**Comment by mpalavrov** [comment]
Hi @pedroigor,
The suggestion for improvement I made (which is something we would like to have), is to have an attribute to be readonly (or respectfully hidden), based on its value (which is read once only from the already set attributes, not reading it from the values provided after submitting the form for registration/updateProfile).
What @ssilvert added is to be able to have attribute to be required (maybe also readOnly/hidden), based on another attribute value, which is very good addition, and would be good to add both at once, so no changes are made on a later stage (I guess so at least).
To summarize (what we need):
- Attribute **userAttributeA** to be readOnly based on annotation, set on itself, for example "readonly-criteria" with value "valueA". If it is valueA (already defined in the user Attribute), the attribute is readonly in the UpdateProfile for the user, if the value is different, it is behaving normally. To accept RegEx as well.

In addition, it can be expanded with making the elements hidden, using another annotation, but what we currently need is "readonly" based on some criteria evaluating the currently set value of the attrubute.

What @ssilvert added is also I think a must to have, being able to make attribute required, based on the value of another attribute. For example, having 2 annotations on "attributeA" like ("attribute-to-evaluate-to-make-required" set to "attributeB" and "value-to-evaluate-to-make-required" set to "DE", if it is for country for example)

**Reply by pedroigor** [reply]
I see now ... So it is about marking as read-only based on the current attribute value (not really when updating it).

Makes sense to me, but do we also want to run server-side validation when the attribute is read-only because of its value? Otherwise, you will be able to easily bypass this by changing the attribute value when submitting forms or using REST.

**Reply by jbman** [reply]
Exactly, the backend should prevent the update once the value matches the regexp criteria. 
It is also needed to pass the read-only state to the template, because the only of stored value matches the regexp, it should be displayed as read-only. While the user is editing, it should be allowed to switch back until the form could be submitted and was accepted without validation errors.

**Reply by pedroigor** [reply]
I see. I would say that using annotations is perhaps beyond its initial goal where they are mainly targeted to help drive how UIs should render attributes.

Looks like we need a first-class setting in the user profile configuration. But I don't know how it would look like ... If we are going to leverage the existing "permissions" settings for an attribute or something else.

In any case, I think we can start creating an issue and add it to this epic https://github.com/keycloak/keycloak/issues/25542. We are using that epic as a wish list to help us plan the follow-up tasks we want to deliver after KC 24. Do you mind creating one?

**Reply by jbman** [reply]
Thanks, I had the same thoughts about a setting directly at the attribute instead of an annotation. Can create an issue for the epic.

**Reply by jbman** [reply]
@pedroigor Added issue https://github.com/keycloak/keycloak/issues/25707. Could you please update the epic to include it there?

**Reply by pedroigor** [reply]
Thanks, @jbman!

**Reply by antikalk** [reply]
I like the proposal. Maybe just one additional thing that might be cool to consider is to allow for different ways of deciding this:
Additionally to a regex pattern check for the attribute value it would be cool when you could potentially provide/implement a custom SPI for different checks, slightly changing the propused configuration of the issue: 

```json
...
"unmodifiable":  {
   "providerId": "attribute-regex-unmodifiable-check",
   "config": {
      "pattern": ".*"
   },
   "unmodifiable_for": ["user", "admin"],
   "hide_from": ["user"]  
},
...
```
